src/Bonami/Collection/Applicative1.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Bonami\Collection;
+
+/** @template T */
+trait Applicative1
+{
+    /**
+     * @template A
+     *
+     * @phpstan-param A $value
+     *
+     * @phpstan-return static<A>
+     */
+    abstract public static function pure($value);
+
+    /**
+     * @template A
+     * @template B
+     *
+     * @param self<callable(A): B> $closure
+     * @param self<A> $argument
+     *
+     * @return self<B>
+     */
+    abstract public static function ap(self $closure, self $argument): self;
```

Comment:
The more I think about it, more I am convinced that (especially for newcomers) its better to define Applicative in terms of `product` function (or some variant of it) -
in ts it'd be roughly like this (for `Option`)
```ts
<A,B>(fa: Option<A>, fb: Option<B>) => Option<[A, B]>
```

see https://typelevel.org/cats/typeclasses/applicative.html#applicative for details

Also it would not need to be static.. 
```php
    /**
     * @template B
     *
     * @param self<B> $fb
     * @return self<array{0: T, 1: B}>
     */
abstract public function butNotSureAboutTheName(self $fb): self;
```

On the other hand, not sure, how comfortable the work with tuples is in php at the moment..